### PR TITLE
Autodetect methods

### PR DESCRIPTION
**Pull request by tonatoz**

### CODE REVIEW

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/UpstreamRpcMethodsDetector.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            .fromIterable(rpcMethods())
+            .flatMap { (method, params) ->
+                upstream
+                    .getIngressReader()
+                    .read(ChainRequest(method, params))
+                    .flatMap(ChainResponse::requireResult)
+                    .map { mapOf(method to true) }
+                    .onErrorResume {
+                        log.warn(
+                            "Can't detect rpc method $method of upstream ${upstream.getId()}, reason - {}",
```
Comment (KirillPamPam) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/UpstreamRpcMethodsDetector.kt:36]: Let's replace `$method` and `${upstream.getId()}` also with placeholder `{}`

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/UpstreamRpcMethodsDetector.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                            it.message,
+                        )
+                        mapOf(method to false).toMono()
+                    }
+            }.toMono()
```
Comment (KirillPamPam) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/UpstreamRpcMethodsDetector.kt:41]: There will be emitted the very first item, so it doesn't work as expected, it must be changed to `Mono.zip`, for example

Comment (KirillPamPam) [reply src/main/kotlin/io/emeraldpay/dshackle/upstream/UpstreamRpcMethodsDetector.kt:41]: We need something like that
```
Mono.zip(
     rpcMethods().map {
         Mono.just(it)
             .flatMap { (method, param) ->
                 upstream.getIngressReader()
                     .read(ChainRequest(method, param))
                     .flatMap(ChainResponse::requireResult)
                     .map { method to true }
                     .onErrorReturn(
                         method to false
                     )
             }
     }
        ) {
            it.map { p -> p as Pair<String, Boolean> }
                .associate { it.first to it.second }
        }
```

Also this code should be covered with tests

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/UpstreamRpcMethodsDetector.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+// Should be Eth network only?
+class BasicEthUpstreamRpcMethodsDetector(
+    upstream: Upstream,
+) : UpstreamRpcMethodsDetector(upstream) {
+    override fun detectByMagicMethod(): Mono<List<String>> = Mono.empty()
+
+    override fun rpcMethods(): Set<Pair<String, CallParams>> = setOf("eth_getBlockReceipts" to ListParams("latest"))
+}
+
+class BasicPolkadotUpstreamRpcMethodsDetector(
```
Comment (KirillPamPam) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/UpstreamRpcMethodsDetector.kt:57]: Let's move this class and above one under their packages - `ethereum` and `polkadot`

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/ethereum/BasicEthUpstreamRpcMethodsDetector.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    private fun parseRpcModules(data: ByteArray): Map<String, Boolean> {
+        val modules = Global.objectMapper.readValue(data, object : TypeReference<HashMap<String, String>>() {})
+        val allDisabledMethods =
+            DefaultEthereumMethods(upstream.getChain())
+                .getSupportedMethods()
+                .filter { method ->
+                    modules.all { (module, _) -> method.startsWith(module).not() }
+                }.associateWith { false }
+        // Don't trust the modules, check the methods from rpcMethods
+        val rpcMethodsResponse = detectByMethod().block() ?: emptyMap()
```
Comment (KirillPamPam) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/ethereum/BasicEthUpstreamRpcMethodsDetector.kt:42]: We shouldn't use `block()` in non-blocking pipeline

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/ethereum/BasicEthUpstreamRpcMethodsDetector.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            "eth_getBlockReceipts" to ListParams("latest"),
+        )
+
+    private fun parseRpcModules(data: ByteArray): Map<String, Boolean> {
+        val modules = Global.objectMapper.readValue(data, object : TypeReference<HashMap<String, String>>() {})
+        val allDisabledMethods =
+            DefaultEthereumMethods(upstream.getChain())
+                .getSupportedMethods()
+                .filter { method ->
+                    modules.all { (module, _) -> method.startsWith(module).not() }
```
Comment (KirillPamPam) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/ethereum/BasicEthUpstreamRpcMethodsDetector.kt:39]: Not sure that I understand this code, could you explain pls?

File: src/test/kotlin/io/emeraldpay/dshackle/upstream/ethereum/BasicEthUpstreamRpcMethodsDetectorTest.kt [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            }
+
+        val upstream =
+            mock<Upstream> {
+                on { getIngressReader() } doReturn reader
+                on { getChain() } doReturn Chain.ETHEREUM__MAINNET
+            }
+        val detector = BasicEthUpstreamRpcMethodsDetector(upstream)
+        val result = detector.detectRpcMethods().block()
+        assert(result != null)
```
Comment (KirillPamPam) [inline src/test/kotlin/io/emeraldpay/dshackle/upstream/ethereum/BasicEthUpstreamRpcMethodsDetectorTest.kt:49]: Let's use `org.assertj.core.api.Assertions` like in other tests

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/ethereum/BasicEthUpstreamRpcMethodsDetector.kt (hunk 5 above)
Comment (tonatoz) [reply src/main/kotlin/io/emeraldpay/dshackle/upstream/ethereum/BasicEthUpstreamRpcMethodsDetector.kt:39]: We iter over all supported methods and for each method check it start with module name (came from rpc_modules like "eth"/"web3" & etc) and if we can't find any prefix from modules for this method - we flag it as disabled

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/UpstreamRpcMethodsDetector.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+                    .flatMap { (method, param) ->
+                        upstream
+                            .getIngressReader()
+                            .read(ChainRequest(method, param))
+                            .flatMap(ChainResponse::requireResult)
+                            .map { method to true }
+                            .onErrorResume { err ->
+                                val notAvailableError =
+                                    listOf(
+                                        "method ([A-Za-z0-9_]+) does not exist/is not available",
```
Comment (KirillPamPam) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/UpstreamRpcMethodsDetector.kt:33]: Let's initialize the list of regexps once. We don't need to do such operation each time we encounter an error

**Review by KirillPamPam — APPROVED** [review]

**Pull request merged**